notebooks/transport/http.go

```diff
@@ -0,0 +1,196 @@
+package transport
+
+import (
+	"net/http"
+	"strconv"
+
+	"github.com/influxdata/influxdb/v2"
+	feature "github.com/influxdata/influxdb/v2/kit/feature"
+	"github.com/influxdata/influxdb/v2/kit/platform"
+	"github.com/influxdata/influxdb/v2/kit/platform/errors"
+	kithttp "github.com/influxdata/influxdb/v2/kit/transport/http"
+
+	"github.com/go-chi/chi"
+	"github.com/go-chi/chi/middleware"
+	"go.uber.org/zap"
+)
+
+// NotebookHandler is the handler for the notebook service
+type NotebookHandler struct {
+	chi.Router
+
+	api *kithttp.API
+	log *zap.Logger
+}
+
+const (
+	prefixNotebooks = "/api/v2private/flows"
+)
+
+func NewNotebookHandler(log *zap.Logger) *NotebookHandler {
+	h := &NotebookHandler{
+		log: log,
+		api: kithttp.NewAPI(kithttp.WithLog(log)),
+	}
+
+	r := chi.NewRouter()
+	r.Use(
+		middleware.Recoverer,
+		middleware.RequestID,
+		middleware.RealIP,
+		h.notebookFlag, // temporary, remove when feature flag for notebooks is removed
+	)
+
+	r.Route("/orgs/{orgID}/flows", func(r chi.Router) {
+		r.Get("/", h.handleGetNotebooks)
+		r.Post("/", h.handleCreateNotebook)
+
+		r.Route("/{id}", func(r chi.Router) {
+			r.Get("/", h.handleGetNotebook)
+			r.Patch("/", h.handlePatchNotebook)
+			r.Delete("/", h.handleDeleteNotebook)
+		})
+	})
+
+	h.Router = r
+
+	return h
+}
+
+func (h *NotebookHandler) Prefix() string {
+	return prefixNotebooks
+}
+
+// notebookFlag is middleware for returning no content if the notebooks feature
+// flag is set to false. remove this middleware when the feature flag is removed.
+func (h *NotebookHandler) notebookFlag(next http.Handler) http.Handler {
+	fn := func(w http.ResponseWriter, r *http.Request) {
+		flags := feature.FlagsFromContext(r.Context())
+
+		if !flags["notebooks"].(bool) {
```

Comment:
Instead of checking this every time, could we check the flag at the top-level of `notebookFlag` and return a different `http.Handler` depending on the result?
```go
if !flags["notebooks"].(bool) {
  return func()...
}
return func()...
```

Reply:
Since I'm getting the `flags` map off the request context which is annotated higher in the stack, the check has to come from within a request, so doing it like that it has to be checked every time. From what I think you're suggesting, there would need to be access to the `flags` map outside of a request.

Something equivalent could be done with some additional of effort, maybe by passing a `flagger` to the `NewNotebookHandler` and getting the flag map there, then using the flag to shut down any requests if the flag is off. This would set the behavior outside of an individual request.

I wasn't inclined to do something like this for performance reasons since this flag check shouldn't be around for very long, and to generate any traffic to the `api/v2private/flows` route you'd have to have the UI show the notebooks stuff which would require manually turning the flag on either on build or when running `influxd`. And since it's just grabbing values from a map, it should be pretty quick even in that case.

That said, it probably took me longer to formulate this reply than it would have to just make the change I described, so definitely let me know if you have other reasons for this or if I'm missing something!